# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [148]:
import pandas as pd
import sqlite3
import warnings
warnings.simplefilter('ignore')

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [149]:
sales_connection = sqlite3.connect('go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [187]:
tables_to_read = ['product', 'product_type', 'product_line', 'sales_staff', 'sales_branch', 
                  'retailer_site', 'country', 'order_header', 'order_details',
                  'returned_item', 'return_reason']

dataframes = {}

for table in tables_to_read:
    query = f"SELECT * FROM {table}"
    dataframes[table] = pd.read_sql_query(query, sales_connection)

product_df = dataframes['product']
product_type_df = dataframes['product_type']
product_line_df = dataframes['product_line']
sales_staff_df = dataframes['sales_staff']
sales_branch_df = dataframes['sales_branch']
retailer_site_df = dataframes['retailer_site']
country_df = dataframes['country']
order_header_df = dataframes['order_header']
order_details_df = dataframes['order_details']
returned_item_df = dataframes['returned_item']
return_reason_df = dataframes['return_reason']

print(sales_staff_df)

    SALES_STAFF_CODE FIRST_NAME   LAST_NAME                   POSITION_EN  \
0                100     Tuomas  Savolainen  Level 2 Sales Representative   
1                101   Chang-ho         Kim  Level 2 Sales Representative   
2                102    Jung-ho        Choi  Level 3 Sales Representative   
3                103  Kwei-Ping         Kao  Level 2 Sales Representative   
4                104  Chin-Tsai        Fang  Level 3 Sales Representative   
..               ...        ...         ...                           ...   
97                95      Tomás    Iglesias  Level 2 Sales Representative   
98                96     Bayard       Lopes                Branch Manager   
99                97     Viviam      Barros  Level 3 Sales Representative   
100               98    Eduardo   Guimarães  Level 2 Sales Representative   
101               99     Helena    Lindholm  Level 3 Sales Representative   

              WORK_PHONE EXTENSION                  FAX  \
0    +358(0)17 -

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [151]:
product_df['PRODUCTION_COST'] = pd.to_numeric(product_df['PRODUCTION_COST'], errors='coerce')
producten_filtered = product_df.loc[(product_df['PRODUCTION_COST'] > 50) & (product_df['PRODUCTION_COST'] < 100), ['PRODUCT_NAME', 'PRODUCTION_COST']]
producten_filtered

PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [152]:
product_df['MARGIN'] = pd.to_numeric(product_df['MARGIN'])
result = product_df[['PRODUCT_NAME', 'MARGIN']][(product_df['MARGIN'] < 0.20) | (product_df['MARGIN'] > 0.60)]
#result

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [153]:
countries_francs = country_df[['COUNTRY', 'CURRENCY_NAME']][country_df['CURRENCY_NAME'] == 'francs']
# countries_francs

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [154]:
product_df['MARGIN'] = pd.to_numeric(product_df['MARGIN'], errors='coerce')
products_high_margin = product_df[['INTRODUCTION_DATE', 'MARGIN']][product_df['MARGIN'] > 0.50]
products_high_margin_unique_dates = products_high_margin.drop_duplicates(subset='INTRODUCTION_DATE')
# products_high_margin_unique_dates

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [155]:
verkoopafdelingen = sales_branch_df[['ADDRESS1', 'CITY']][(sales_branch_df['ADDRESS2'].notna()) & (sales_branch_df['REGION'].notna())]
# verkoopafdelingen

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [166]:
merged_df = pd.merge(product_df, product_type_df, on="PRODUCT_TYPE_CODE")
product_counts = merged_df.groupby("PRODUCT_TYPE_CODE").size()
product_type_counts_df = pd.DataFrame({"TYPE_CODE": product_counts.index, "AANTAL_PRODUCTEN": product_counts.values})
product_type_counts_df


TYPE_CODE  AANTAL_PRODUCTEN
0          1                10
1         10                 5
2         11                 5
3         12                 5
4         13                 4
5         14                 4
6         15                 5
7         16                 5
8         17                 5
9         18                 4
10        19                 4
11         2                 6
12        20                 3
13        21                 4
14         3                 7
15         4                 6
16         5                12
17         6                 4
18         7                 4
19         8                 7
20         9                 6

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [173]:
merged_df = pd.merge(sales_branch_df, country_df, on="COUNTRY_CODE")
vestigingen_land = merged_df.groupby("COUNTRY_CODE").size()
vestigingen_land_df = pd.DataFrame({"Country_Code": vestigingen_land.index, "Aantal_Vestigingen":vestigingen_land.values})
vestigingen_land_df

Country_Code  Aantal_Vestigingen
0             1                   2
1            10                   1
2            11                   2
3            12                   1
4            13                   1
5            14                   1
6            16                   1
7            17                   1
8            19                   1
9             2                   2
10           20                   1
11           21                   1
12           22                   1
13            3                   4
14            4                   2
15            5                   1
16            6                   1
17            7                   1
18            8                   1
19            9                   2

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [184]:
cooking_gear_code = product_type_df[product_type_df["PRODUCT_TYPE_EN"] == "Cooking Gear"]["PRODUCT_TYPE_CODE"]
product_names = product_df[product_df['PRODUCT_TYPE_CODE'].isin([cooking_gear_code])]['PRODUCT_NAME']

Series([], Name: PRODUCT_NAME, dtype: object)

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [157]:
# Find all sales staff codes in order_header_df
existing_codes = order_header_df['SALES_STAFF_CODE']

# Filter sales_staff_df to exclude codes in existing_codes
staff_without_sales = sales_staff_df[~sales_staff_df['SALES_STAFF_CODE'].isin(existing_codes)]
selected_columns = ['FIRST_NAME', 'LAST_NAME']
filtered_staff = staff_without_sales[selected_columns]

# Print the DataFrame
#filtered_staff

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [158]:
numericMargin = pd.to_numeric(product_df['MARGIN'])
total_margin = numericMargin.sum()
# shape wordt gebruikt om het aantal rijen en kolommen te bepalen door[0] bepaal je aantal rijen en met [1] aantal kolommen
total_Products = product_df['PRODUCT_NUMBER'].count()
avg_margin_product = total_margin/total_Products
low_margin_products = product_df['MARGIN'] < avg_margin_product

total_low_margin = low_margin_products.sum()
data = {
    "Average Margin": [avg_margin_product],
    "Products Below Avg": [low_margin_products.sum()]
}

overview_df = pd.DataFrame(data)

# Print the DataFrame
print(overview_df)

   Average Margin  Products Below Avg
0        0.396609                  59


Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [159]:
products_over500 = product_df["PRODUCTION_COST"] * (1 +product_df["MARGIN"]) > 500
over500_products_df = product_df[products_over500]
over500_product_names = over500_products_df["PRODUCT_NAME"]
over500_product_names #klopt niet met de 13 rijen die verwacht worden en ik moet ook nog de check uitvoeren of ze nog nooit zijn teruggebracht


4              Hailstorm Titanium Irons
6         Lady Hailstorm Titanium Irons
7          Hailstorm Titanium Woods Set
8             Hailstorm Steel Woods Set
9     Lady Hailstorm Titanium Woods Set
10       Lady Hailstorm Steel Woods Set
19                            Star Dome
21                         Star Gazer 3
22                         Star Gazer 6
55                       Husky Rope 200
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [188]:
naam_functie = sales_staff_df['LAST_NAME'], sales_staff_df['POSITION_EN']
naam_functie

(0      Savolainen
 1             Kim
 2            Choi
 3             Kao
 4            Fang
           ...    
 97       Iglesias
 98          Lopes
 99         Barros
 100     Guimarães
 101      Lindholm
 Name: LAST_NAME, Length: 102, dtype: object,
 0      Level 2 Sales Representative
 1      Level 2 Sales Representative
 2      Level 3 Sales Representative
 3      Level 2 Sales Representative
 4      Level 3 Sales Representative
                    ...             
 97     Level 2 Sales Representative
 98                   Branch Manager
 99     Level 3 Sales Representative
 100    Level 2 Sales Representative
 101    Level 3 Sales Representative
 Name: POSITION_EN, Length: 102, dtype: object)

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [161]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [162]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 